packages/gatsby-adapter-netlify/src/pretty-urls.ts

```diff
@@ -0,0 +1,90 @@
+import fastq from "fastq"
+import fs from "fs-extra"
+
+export function generatePrettyUrlFilePath(routePath: string): string {
+  if (routePath.endsWith(`/`)) {
+    return `${routePath}index.html`
+  } else {
+    return `${routePath}.html`
+  }
+}
+
+interface IMoveTask {
+  from: string
+  to: string
+  keepOriginalFile: boolean
+}
+
+export function createStaticAssetsPathHandler(): {
+  ensureStaticAssetPath: (
+    filePath: string,
+    routePath: string
+  ) => { finalFilePath: string; isDynamic: boolean }
+  fileMovingDone: () => Promise<void>
+} {
+  const moveQueue = fastq<void, IMoveTask, void>(async (task, cb) => {
+    try {
+      if (task.keepOriginalFile) {
+        await fs.copy(task.from, task.to, { overwrite: true })
+      } else {
+        await fs.move(task.from, task.to, { overwrite: true })
+      }
+      cb(null, undefined)
+    } catch (error) {
+      cb(error)
+    }
+  }, 2)
+
+  function ensureStaticAssetPath(
+    filePath: string,
+    routePath: string
+  ): { finalFilePath: string; isDynamic: boolean } {
+    const shouldUsePrettyUrl =
+      filePath.endsWith(`.html`) && !routePath.endsWith(`.html`)
+
+    let isDynamic = false
+    // dynamic routes syntax use characters that are reserved in a lot of filesystems
+    // so if route is dynamic we should normalize filepath
+    if (routePath.includes(`:`) || routePath.includes(`*`)) {
+      routePath = routePath
+        // replace `:param` with `[param]`
+        .replace(/:([^:/\\]+)/gm, `[$1]`)
+        // replace `*param` with `[...param]` and `*` with `[...]`
+        .replace(/\*([^:/\\]*|$)/gm, `[...$1]`)
```

Comment:
Might be nice to pull out this normalization into its own function.

Reply:
moved in https://github.com/gatsbyjs/gatsby/pull/38666/commits/9e9115a3b15b312422963907d1b1c14c16f3762a also added some basic tests just for that normalization ... and "fixed" one of existing tests that wasn't really testing what it said it was testing